pilot/pkg/serviceregistry/kube/controller/ambient/sidecar_interop.go

```diff
@@ -0,0 +1,139 @@
+	"istio.io/istio/pkg/ptr"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/workloadapi"
+)
+
+// serviceEDS represents a service and a list of all instances of waypoints.
+// For example, we may have ServiceKey=httpbin.org and WaypointInstance=[list of *waypoint workloads* attached].
+// This is to map to the eventual EDS structure.
+type serviceEDS struct {
+	ServiceKey       string
+	WaypointInstance []*workloadapi.Workload
+}
+
+func (w serviceEDS) ResourceName() string {
+	return w.ServiceKey
+}
+
+// RegisterEdsShim handles triggering xDS events when Envoy EDS needs to change.
+// Most of ambient index works to build `workloadapi` types - Workload, Service, etc.
+// Envoy uses a different API, with different relationships between types.
+// To ensure Envoy are updated properly on changes, we compute this information.
+// Currently, this is only used to trigger events.
+// Ideally, the information we are using in Envoy and the event trigger are using the same data directly.
+func RegisterEdsShim(
+	xdsUpdater model.XDSUpdater,
+	Workloads krt.Collection[model.WorkloadInfo],
+	WorkloadsByServiceKey krt.Index[string, model.WorkloadInfo],
+	Services krt.Collection[model.ServiceInfo],
+	ServicesByAddress krt.Index[networkAddress, model.ServiceInfo],
+) {
+	ServiceEds := krt.NewCollection(
+		Services,
+		func(ctx krt.HandlerContext, svc model.ServiceInfo) *serviceEDS {
+			if !svc.Waypoint.IngressUseWaypoint {
+				// Currently, we only need this for ingres -> waypoint usage
+				// If we extend this to sidecars, etc we can drop this.
+				return nil
+			}
+			wp := svc.Service.Waypoint
+			if wp == nil {
+				return nil
+			}
+			var waypointServiceKey string
+			switch addr := wp.Destination.(type) {
+			// Easy case: waypoint is already a hostname. Just return it directly
+			case *workloadapi.GatewayAddress_Hostname:
+				hn := addr.Hostname
+				waypointServiceKey = hn.Namespace + "/" + hn.Hostname
+			// Hard case: waypoint is an IP address. Need to look it up.
+			case *workloadapi.GatewayAddress_Address:
+				wAddress := addr.Address
+				serviceKey := networkAddress{
+					network: wAddress.Network,
+					ip:      mustByteIPToString(wAddress.Address),
+				}
+				waypointSvc := krt.FetchOne(ctx, Services, krt.FilterIndex(ServicesByAddress, serviceKey))
+				if waypointSvc == nil {
+					return nil
+				}
+				waypointServiceKey = waypointSvc.ResourceName()
+			}
+			workloads := krt.Fetch(ctx, Workloads, krt.FilterIndex(WorkloadsByServiceKey, waypointServiceKey))
+			return &serviceEDS{
+				ServiceKey: svc.ResourceName(),
+				WaypointInstance: slices.Map(workloads, func(e model.WorkloadInfo) *workloadapi.Workload {
+					return e.Workload
+				}),
+			}
+		},
+		krt.WithName("ServiceEds"))
+	ServiceEds.RegisterBatch(
+		PushXds(xdsUpdater, func(svc serviceEDS) model.ConfigKey {
+			ns, hostname, _ := strings.Cut(svc.ServiceKey, "/")
+			return model.ConfigKey{Kind: kind.ServiceEntry, Name: hostname, Namespace: ns}
```

Comment:
can you make this trigger push only for gateway?

Reply:
I don't think we have that capability in the push system. Maybe we can enhance it in the future?